cranelift-codegen/src/isa/x86/enc_tables.rs

```diff
@@ -895,6 +895,80 @@ fn expand_fcvt_to_uint_sat(
     cfg.recompute_ebb(pos.func, done);
 }
 
+/// Convert shuffle instructions.
+fn convert_shuffle(
+    inst: ir::Inst,
+    func: &mut ir::Function,
+    _cfg: &mut ControlFlowGraph,
+    _isa: &dyn TargetIsa,
+) {
+    let mut pos = FuncCursor::new(func).at_inst(inst);
+    pos.use_srcloc(inst);
+
+    if let ir::InstructionData::Shuffle { args, mask, .. } = pos.func.dfg[inst] {
+        // A mask-building helper: in 128-bit SIMD, 0-15 indicate which lane to read from and a 1
+        // in the most significant position zeroes the lane.
+        let zero_unknown_lane_index = |b: u8| if b > 15 { 0b10000000 } else { b };
+
+        // We only have to worry about aliasing here because copies will be introduced later (in
+        // regalloc).
+        let a = pos.func.dfg.resolve_aliases(args[0]);
+        let b = pos.func.dfg.resolve_aliases(args[1]);
+        let mask = pos
+            .func
+            .dfg
+            .immediates
+            .get(mask)
+            .expect("The shuffle immediate should have been recorded before this point")
+            .clone();
```

Comment:
Out of curiosity, do we need the clone here, or can we just use a mask slice thereafter?

Reply:
Well, we need it unless we change other stuff--Rust complains that we can't borrow too much from `pos`:

```
error[E0502]: cannot borrow `pos` as mutable because it is also borrowed as immutable
   --> cranelift-codegen/src/isa/x86/enc_tables.rs:943:30
    |
917 |           let mask = pos
    |  ____________________-
918 | |             .func
919 | |             .dfg
920 | |             .immediates
    | |_______________________- immutable borrow occurs here
...
943 |               let mask_value = pos.ins().vconst(a_type, handle);
    |                                ^^^^^^^^^ mutable borrow occurs here
...
948 |               let constructed_mask = mask
    |                                      ---- immutable borrow later used here
```